ssl/ech.c

```diff
@@ -0,0 +1,5905 @@
+        WPACKET_cleanup(&hpkt);
+        BUF_MEM_free(hpkt_mem);
+        return 0;
+    }
+    WPACKET_get_total_written(&hpkt, toklen);
+    *hrrtok = OPENSSL_malloc(*toklen);
+    if (*hrrtok == NULL)
+        return 0;
+    memcpy(*hrrtok, WPACKET_get_curr(&hpkt) - *toklen, *toklen);
+    WPACKET_cleanup(&hpkt);
+    BUF_MEM_free(hpkt_mem);
+    return 1;
+}
+
+/* SECTION: Non-public functions used elsewhere in the library */
+
+# ifdef OSSL_ECH_SUPERVERBOSE
+/*
+ * @brief ascii-hex print a buffer nicely for debug/interop purposes
+ * @param msg pre-pend to trace lines
+ * @param buf points to the buffer to print
+ * @param blen is the length of buffer to print
+ */
+void ech_pbuf(const char *msg, const unsigned char *buf, const size_t blen)
+{
+    OSSL_TRACE_BEGIN(TLS) {
+        if (msg == NULL) {
+            BIO_printf(trc_out, "msg is NULL\n");
+        } else if (buf == NULL || blen == 0) {
+            BIO_printf(trc_out, "%s: buf is %p\n", msg, (void *)buf);
+            BIO_printf(trc_out, "%s: blen is %lu\n", msg, (unsigned long)blen);
+        } else {
+            size_t i;
+
+            BIO_printf(trc_out, "%s (%lu):\n    ", msg, (unsigned long)blen);
+            for (i = 0; i < blen; i++) {
+                if ((i != 0) && (i % 16 == 0))
+                    BIO_printf(trc_out, "\n    ");
+                BIO_printf(trc_out, "%02x:", (unsigned)(buf[i]));
+            }
+            BIO_printf(trc_out, "\n");
+        }
+    } OSSL_TRACE_END(TLS);
+    return;
+}
+
+/*
+ * @brief trace out transcript
+ * @param msg pre-pend to trace lines
+ * @param s is the SSL connection
+ */
+void ech_ptranscript(const char *msg, SSL_CONNECTION *s)
+{
+    size_t hdatalen = 0;
+    unsigned char *hdata = NULL;
+    unsigned char ddata[1000];
+    size_t ddatalen;
+
+    if (s == NULL)
+        return;
+    hdatalen = BIO_get_mem_data(s->s3.handshake_buffer, &hdata);
+    ech_pbuf(msg, hdata, hdatalen);
+    if (s->s3.handshake_dgst != NULL) {
+        if (ssl_handshake_hash(s, ddata, 1000, &ddatalen) == 0) {
+            OSSL_TRACE_BEGIN(TLS) {
+                BIO_printf(trc_out, "ssl_handshake_hash failed\n");
+            } OSSL_TRACE_END(TLS);
+        }
+        ech_pbuf(msg, ddata, ddatalen);
+    } else {
+        OSSL_TRACE_BEGIN(TLS) {
+            BIO_printf(trc_out, "handshake_dgst is NULL\n");
+        } OSSL_TRACE_END(TLS);
+    }
+    return;
+}
+# endif
+
+/*
+ * @brief Free an ECHConfig structure's internals
+ * @param tbf is the thing to be freed
+ */
+void ECHConfig_free(ECHConfig *tbf)
+{
+    unsigned int i = 0;
+
+    if (tbf == NULL)
+        return;
+    OPENSSL_free(tbf->public_name);
+    OPENSSL_free(tbf->pub);
+    OPENSSL_free(tbf->ciphersuites);
+    OPENSSL_free(tbf->exttypes);
+    OPENSSL_free(tbf->extlens);
+    for (i = 0; i != tbf->nexts; i++)
+        OPENSSL_free(tbf->exts[i]);
+    OPENSSL_free(tbf->exts);
+    OPENSSL_free(tbf->encoding_start);
+    memset(tbf, 0, sizeof(ECHConfig));
```

Comment:
The ``memset()`` isn't really needed no, but did help debugging earlier.
As to why not free the passed object, it's because there can be an array of 'em, so the higher level code does that.